presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Pattern.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.iterative;
+
+import com.facebook.presto.sql.planner.plan.PlanNode;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public abstract class Pattern
+{
+    private Pattern() {}
+
+    public abstract boolean matches(PlanNode node);
+
+    public static Pattern any()
+    {
+        return AnyNode.INSTANCE;
+    }
+
+    public static <T extends PlanNode> Pattern node(Class<T> nodeClass)
+    {
+        return new ClassNodeBased(nodeClass);
+    }
+
+    static class AnyNode
+            extends Pattern
+    {
+        static final AnyNode INSTANCE = new AnyNode();
+
+        @Override
+        public boolean matches(PlanNode node)
+        {
+            return true;
+        }
+    }
+
+    static class ClassNodeBased<T extends PlanNode>
```

Comment:
I don't think I like this name. Maybe something like "MatchNodeClass" or "MatchNodeType"